neuro_san/service/http/server/http_sidecar.py

```diff
@@ -144,34 +144,64 @@ def make_app(self, requests_limit: int, logger: EventLoopLogger):
+        """
+        Quick low-impact check to see if the allowed list of agents need updating.
+
+        :return: True if the allowed_agents list needs updating. False otherwise
+        """
+        # Start by taking a snapshot of when we were last modified without the lock
+        last_modified: float = self.allowed_last_modified
+
+        for network_storage in self.network_storage_dict.values():
+            if last_modified < network_storage.get_last_modified():
+                # Something has changed since the last time we looked
+                return True
+
+        # Nothing has changed
+        return False
 
     def update_agents(self, metadata: Dict[str, Any]):
         """
         Update list of agents for which serving is allowed.
         :param metadata: metadata to be used for logging if necessary.
         :return: nothing
         """
-        data: Dict[str, Any] = {}
+        _ = metadata
+
+        # First do a quick check against network storage timestamps to see if
+        # we need to anything more heavyweight.
+        if not self.is_needing_allowed_update():
+            return
+
+        # Keep a list of the new allowed agents.
+        new_allowed_names: List[str] = []
+        for network_storage in self.network_storage_dict.values():
+            # Add to the list of agents from the storage as the new allowed set.
+            agent_names: List[str] = network_storage.get_agent_names()
+            new_allowed_names.extend(agent_names)
 
-        public_storage: AgentNetworkStorage = self.network_storage_dict.get("public")
-        # Why do we need the Concierge if we already have access to public_storage?
-        session: ConciergeSession = DirectConciergeSession(public_storage, metadata=metadata)
-        agents_dict: Dict[str, List[Dict[str, str]]] = session.list(data)
-        agents_list: List[Dict[str, str]] = agents_dict["agents"]
-        agents: List[str] = []
-        for agent_dict in agents_list:
-            agents.append(agent_dict["agent_name"])
         with self.lock:
-            # We assume all agents from "agents" list are enabled:
-            for agent_name in agents:
-                if self.allowed_agents.get(agent_name, None) is None:
+
+            modified: bool = False
+
+            # We assume all agents from "new_allowed_names" list are enabled:
+            for agent_name in new_allowed_names:
+                if self.allowed_agents.get(agent_name) is None:
                     self.add_agent(agent_name)
+                    modified = True
+
             # All other agents are disabled:
             allowed_set = set(self.allowed_agents.keys())
             for agent_name in allowed_set:
-                if agent_name not in agents:
+                if agent_name not in new_allowed_names:
                     self.remove_agent(agent_name)
+                    modified = True
+
+            if modified:
+                self.allowed_last_modified = time.time()
```

Comment:
Only set the timestamp if something changed.
Theoretically this should always happen, but it's not actually obvious that it does.